File: backend/internal/auth/google.go
package auth

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/jackc/pgx/v5/pgconn"
	"github.com/jackc/pgx/v5/pgxpool"
	"practiq/internal/api"
)

// Package-level endpoints so tests can point them at httptest servers.
var (
	googleAuthURL      = "https://accounts.google.com/o/oauth2/v2/auth"
	googleTokenURL     = "https://oauth2.googleapis.com/token"
	googleUserinfoURL  = "https://openidconnect.googleapis.com/v1/userinfo"
	googleTokeninfoURL = "https://oauth2.googleapis.com/tokeninfo"
)

var googleHTTPClient = &http.Client{Timeout: 10 * time.Second}

const googleStateCookieName = "google_oauth_state"

type GoogleClaims struct {
	Sub           string
	Email         string
	EmailVerified bool
}

type googleTokenRequest struct {
	IDToken string `json:"idToken"`
}

func googleClientID() string {
	return strings.TrimSpace(os.Getenv("GOOGLE_CLIENT_ID"))
}

func googleRedirectURL() string {
	return strings.TrimRight(strings.TrimSpace(os.Getenv("APP_ORIGIN")), "/") + "/api/v1/auth/google/callback"
}

func googleAuthDisabled() *api.Error {
	return api.NewError(http.StatusNotFound, "GOOGLE_AUTH_DISABLED", "Google sign-in is not configured", nil)
}

func googleAuthFailed() *api.Error {
	return api.NewError(http.StatusUnauthorized, "GOOGLE_AUTH_FAILED", "Google sign-in failed", nil)
}

// GoogleStart begins the browser authorization-code flow.
func GoogleStart() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		clientID := googleClientID()
		if clientID == "" {
			api.HandleError(w, r, googleAuthDisabled())
			return
		}
		state := NewUUID()
		http.SetCookie(w, &http.Cookie{
			Name:     googleStateCookieName,
			Value:    state,
			Path:     "/api/v1/auth/google",
			MaxAge:   600,
			HttpOnly: true,
			SameSite: http.SameSiteLaxMode,
			Secure:   os.Getenv("NODE_ENV") == "production",
		})
		params := url.Values{
			"client_id":     {clientID},
			"redirect_uri":  {googleRedirectURL()},
			"response_type": {"code"},
			"scope":         {"openid email profile"},
			"state":         {state},
		}
		http.Redirect(w, r, googleAuthURL+"?"+params.Encode(), http.StatusFound)
	})
}

// GoogleCallback finishes the browser flow: code exchange, userinfo, session cookie.
func GoogleCallback(deps HandlerDependencies) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		clientID := googleClientID()
		if clientID == "" {
			api.HandleError(w, r, googleAuthDisabled())
			return
		}
		expireStateCookie(w)
		stateCookie, err := r.Cookie(googleStateCookieName)
		state := r.URL.Query().Get("state")
		if err != nil || state == "" || stateCookie.Value != state {
			api.HandleError(w, r, googleAuthFailed())
			return
		}
		code := r.URL.Query().Get("code")
		if code == "" {
			api.HandleError(w, r, googleAuthFailed())
			return
		}
		claims, err := exchangeGoogleCode(r.Context(), clientID, code)
		if err != nil {
			api.HandleError(w, r, err)
			return
		}
		user, err := deps.GoogleUser(r.Context(), claims)
		if err != nil {
			api.HandleError(w, r, err)
			return
		}
		if err := deps.SetSession(w, user.ID); err != nil {
			api.HandleError(w, r, err)
			return
		}
		target := strings.TrimSpace(os.Getenv("APP_ORIGIN"))
		if target == "" {
			target = "/"
		}
		http.Redirect(w, r, target, http.StatusFound)
	})
}

// GoogleToken authenticates mobile clients that obtained a Google ID token natively.
func GoogleToken(deps HandlerDependencies) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if googleClientID() == "" {
			api.HandleError(w, r, googleAuthDisabled())
			return
		}
		var body googleTokenRequest
		if err := decodeAuthRequest(w, r, &body); err != nil {
			api.HandleError(w, r, err)
			return
		}
		if strings.TrimSpace(body.IDToken) == "" {
			api.HandleError(w, r, api.ValidationError([]api.ValidationDetail{{Field: "idToken", Message: "is required"}}))
			return
		}
		if err := rateLimitAuth(r, ""); err != nil {
			api.HandleError(w, r, err)
			return
		}
		claims, err := verifyGoogleIDToken(r.Context(), body.IDToken)
		if err != nil {
			api.HandleError(w, r, err)
			return
		}
		user, err := deps.GoogleUser(r.Context(), claims)
		if err != nil {
			api.HandleError(w, r, err)
			return
		}
		if err := deps.SetSession(w, user.ID); err != nil {
			api.HandleError(w, r, err)
			return
		}
		api.OK(w, r, issuedSessionResponse(w, user), nil)
	})
}

func expireStateCookie(w http.ResponseWriter) {
	http.SetCookie(w, &http.Cookie{
		Name:     googleStateCookieName,
		Value:    "",
		Path:     "/api/v1/auth/google",
		MaxAge:   -1,
		HttpOnly: true,
		SameSite: http.SameSiteLaxMode,
	})
}

func exchangeGoogleCode(ctx context.Context, clientID string, code string) (GoogleClaims, error) {
	form := url.Values{
		"code":          {code},
		"client_id":     {clientID},
		"client_secret": {os.Getenv("GOOGLE_CLIENT_SECRET")},
		"redirect_uri":  {googleRedirectURL()},
		"grant_type":    {"authorization_code"},
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, googleTokenURL, strings.NewReader(form.Encode()))
	if err != nil {
		return GoogleClaims{}, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	resp, err := googleHTTPClient.Do(req)
	if err != nil {
		return GoogleClaims{}, googleAuthFailed()
	}
	defer resp.Body.Close()
	var token struct {
		AccessToken string `json:"access_token"`
	}
	if resp.StatusCode != http.StatusOK || json.NewDecoder(resp.Body).Decode(&token) != nil || token.AccessToken == "" {
		return GoogleClaims{}, googleAuthFailed()
	}

	userinfoReq, err := http.NewRequestWithContext(ctx, http.MethodGet, googleUserinfoURL, nil)
	if err != nil {
		return GoogleClaims{}, err
	}
	userinfoReq.Header.Set("Authorization", "Bearer "+token.AccessToken)
	userinfoResp, err := googleHTTPClient.Do(userinfoReq)
	if err != nil {
		return GoogleClaims{}, googleAuthFailed()
	}
	defer userinfoResp.Body.Close()
	var userinfo struct {
		Sub           string `json:"sub"`
		Email         string `json:"email"`
		EmailVerified bool   `json:"email_verified"`
	}
	if userinfoResp.StatusCode != http.StatusOK || json.NewDecoder(userinfoResp.Body).Decode(&userinfo) != nil || userinfo.Sub == "" {
		return GoogleClaims{}, googleAuthFailed()
	}
	return GoogleClaims{Sub: userinfo.Sub, Email: userinfo.Email, EmailVerified: userinfo.EmailVerified}, nil
}

// verifyGoogleIDToken lets Google's tokeninfo endpoint validate the signature.
// ponytail: one remote call per login; switch to local JWKS verification if volume matters.
func verifyGoogleIDToken(ctx context.Context, idToken string) (GoogleClaims, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, googleTokeninfoURL+"?id_token="+url.QueryEscape(idToken), nil)
	if err != nil {
		return GoogleClaims{}, err
	}
	resp, err := googleHTTPClient.Do(req)
	if err != nil {
		return GoogleClaims{}, googleAuthFailed()
	}
	defer resp.Body.Close()
	var info struct {
		Aud           string `json:"aud"`
		Iss           string `json:"iss"`
		Sub           string `json:"sub"`
		Email         string `json:"email"`
		EmailVerified string `json:"email_verified"`
		Exp           string `json:"exp"`
	}
	if resp.StatusCode != http.StatusOK || json.NewDecoder(resp.Body).Decode(&info) != nil {
		return GoogleClaims{}, googleAuthFailed()
	}
	if info.Sub == "" || (info.Iss != "accounts.google.com" && info.Iss != "https://accounts.google.com") {
		return GoogleClaims{}, googleAuthFailed()
	}
	if !googleAudienceAllowed(info.Aud) {
		return GoogleClaims{}, googleAuthFailed()
	}
	if exp, err := strconv.ParseInt(info.Exp, 10, 64); err != nil || !time.Unix(exp, 0).After(time.Now()) {
		return GoogleClaims{}, googleAuthFailed()
	}
	return GoogleClaims{Sub: info.Sub, Email: info.Email, EmailVerified: info.EmailVerified == "true"}, nil
}

func googleAudienceAllowed(aud string) bool {
	if aud == "" {
		return false
	}
	if aud == googleClientID() {
		return true
	}
	for _, allowed := range strings.Split(os.Getenv("GOOGLE_MOBILE_CLIENT_IDS"), ",") {
		if trimmed := strings.TrimSpace(allowed); trimmed != "" && trimmed == aud {
			return true
		}
	}
	return false
}

// FindOrCreateGoogleUser resolves the account for verified Google claims:
// match by google_sub, else link by verified email, else create a new user.
func FindOrCreateGoogleUser(ctx context.Context, pool *pgxpool.Pool, claims GoogleClaims) (*User, error) {
	user, err := findOrCreateGoogleUserOnce(ctx, pool, claims)
	if isUniqueViolation(err) {
		// Retry once: covers a concurrent first login and username-suffix collisions.
		user, err = findOrCreateGoogleUserOnce(ctx, pool, claims)
	}
	return user, err
}

func findOrCreateGoogleUserOnce(ctx context.Context, pool *pgxpool.Pool, claims GoogleClaims) (*User, error) {
	if claims.Sub == "" {
		return nil, googleAuthFailed()
	}
	row := pool.QueryRow(ctx, `
		SELECT id, username, email, is_active, role, membership
		FROM users
		WHERE google_sub = $1
		LIMIT 1
	`, claims.Sub)
	user, err := scanUser(row)
	if err != nil {
		return nil, err
	}
	if user == nil && claims.Email != "" && claims.EmailVerified {
		row = pool.QueryRow(ctx, `
			UPDATE users
			SET google_sub = $1
			WHERE lower(email) = lower($2) AND google_sub IS NULL
			RETURNING id, username, email, is_active, role, membership
		`, claims.Sub, claims.Email)
		user, err = scanUser(row)
		if err != nil {
			return nil, err
		}
		if user != nil {
			invalidateUserCache(ctx, user.ID)
		}
	}
	if user == nil {
		if claims.Email == "" || !claims.EmailVerified {
			return nil, api.NewError(http.StatusForbidden, "GOOGLE_EMAIL_UNVERIFIED", "Google account email must be verified", nil)
		}
		row = pool.QueryRow(ctx, `
			INSERT INTO users (username, email, google_sub, role, membership)
			VALUES ($1, $2, $3, 'user', 'free')
			RETURNING id, username, email, is_active, role, membership
		`, googleUsername(claims.Email), claims.Email, claims.Sub)
		user, err = scanUser(row)
		if err != nil {
			return nil, err
		}
	}
	if user == nil {
		return nil, googleAuthFailed()
	}
	if !user.IsActive {
		return nil, api.NewError(http.StatusForbidden, "USER_INACTIVE", "User account is disabled", nil)
	}
	cacheUser(ctx, *user)
	return user, nil
}

func isUniqueViolation(err error) bool {
	var pgErr *pgconn.PgError
	return errors.As(err, &pgErr) && pgErr.Code == "23505"
}

func googleUsername(email string) string {
	local := strings.SplitN(email, "@", 2)[0]
	var b strings.Builder
	for _, c := range local {
		if b.Len() >= 12 {
			break
		}
		if c == '_' || (c >= '0' && c <= '9') || (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') {
			b.WriteRune(c)
		}
	}
	name := b.String()
	if name == "" {
		name = "user"
	}
	return name + "_" + strings.ReplaceAll(NewUUID(), "-", "")[:6]
}

File: server/auth/google.py
"""Google OAuth: browser authorization-code flow + mobile ID-token verification."""

import asyncio
import base64
import hashlib
import hmac
import json
import os
import secrets
from dataclasses import dataclass
from urllib.parse import urlencode

import httpx
from fastapi import Request, Response
from google.auth.transport import requests as google_requests
from google.oauth2 import id_token as google_id_token
from psycopg import errors as pg_errors
from psycopg_pool import AsyncConnectionPool

from .. import envelope, redisx
from . import runtime

GOOGLE_AUTH_URL = 'https://accounts.google.com/o/oauth2/v2/auth'
GOOGLE_TOKEN_URL = 'https://oauth2.googleapis.com/token'
GOOGLE_STATE_COOKIE_NAME = 'google_oauth_state'
GOOGLE_OAUTH_TTL_SECONDS = 600

_http_client: httpx.AsyncClient | None = None
_google_request = google_requests.Request()


def _http() -> httpx.AsyncClient:
    global _http_client
    if _http_client is None:
        _http_client = httpx.AsyncClient(timeout=10.0)
    return _http_client


def _google_client_id() -> str:
    return os.environ.get('GOOGLE_CLIENT_ID', '').strip()


def _google_redirect_url() -> str:
    return os.environ.get('APP_ORIGIN', '').strip().rstrip('/') + '/api/v1/auth/google/callback'


def _google_auth_disabled() -> envelope.APIError:
    return envelope.new_error(404, 'GOOGLE_AUTH_DISABLED', 'Google sign-in is not configured')


def _google_auth_failed() -> envelope.APIError:
    return envelope.new_error(401, 'GOOGLE_AUTH_FAILED', 'Google sign-in failed')


@dataclass(frozen=True)
class GoogleClaims:
    sub: str
    email: str
    email_verified: bool


def set_state_cookie(response: Response, state: str) -> None:
    response.set_cookie(
        key=GOOGLE_STATE_COOKIE_NAME,
        value=state,
        path='/api/v1/auth/google',
        max_age=600,
        httponly=True,
        samesite='lax',
        secure=os.environ.get('NODE_ENV') == 'production',
    )


def _random_token() -> str:
    return secrets.token_urlsafe(32)


def _pkce_challenge(verifier: str) -> str:
    return base64.urlsafe_b64encode(hashlib.sha256(verifier.encode()).digest()).rstrip(b'=').decode()


def _oauth_state_key(state: str) -> str:
    return redisx.redis_key('oauth', 'google', hashlib.sha256(state.encode()).hexdigest())


async def google_start_params() -> tuple[str, str]:
    client_id = _google_client_id()
    if not client_id:
        raise _google_auth_disabled()
    state, nonce, code_verifier = _random_token(), _random_token(), _random_token()
    rdb = redisx.client()
    if rdb is None:
        raise envelope.new_error(503, 'OAUTH_STATE_UNAVAILABLE', 'Authentication service is temporarily unavailable')
    try:
        await rdb.set(
            _oauth_state_key(state),
            json.dumps({'nonce': nonce, 'code_verifier': code_verifier}),
            ex=GOOGLE_OAUTH_TTL_SECONDS,
        )
    except Exception as exc:
        raise envelope.new_error(503, 'OAUTH_STATE_UNAVAILABLE', 'Authentication service is temporarily unavailable') from exc
    params = urlencode({
        'client_id': client_id,
        'redirect_uri': _google_redirect_url(),
        'response_type': 'code',
        'scope': 'openid email profile',
        'state': state,
        'nonce': nonce,
        'code_challenge': _pkce_challenge(code_verifier),
        'code_challenge_method': 'S256',
    })
    return f'{GOOGLE_AUTH_URL}?{params}', state


def _expire_state_cookie(response: Response) -> None:
    response.set_cookie(
        key=GOOGLE_STATE_COOKIE_NAME,
        value='',
        path='/api/v1/auth/google',
        max_age=-1,
        httponly=True,
        samesite='lax',
        secure=os.environ.get('NODE_ENV') == 'production',
    )


async def google_callback(request: Request, response: Response, pool: AsyncConnectionPool) -> runtime.User:
    client_id = _google_client_id()
    if not client_id:
        raise _google_auth_disabled()
    _expire_state_cookie(response)
    state_cookie = request.cookies.get(GOOGLE_STATE_COOKIE_NAME, '')
    state = request.query_params.get('state', '')
    if not state_cookie or not state or not hmac.compare_digest(state_cookie, state):
        raise _google_auth_failed()
    code = request.query_params.get('code', '')
    if not code:
        raise _google_auth_failed()
    transaction = await _consume_oauth_transaction(state)
    claims = await _exchange_google_code(client_id, code, transaction['code_verifier'], transaction['nonce'])
    return await find_or_create_google_user(pool, claims)


async def _consume_oauth_transaction(state: str) -> dict[str, str]:
    rdb = redisx.client()
    if rdb is None:
        raise envelope.new_error(503, 'OAUTH_STATE_UNAVAILABLE', 'Authentication service is temporarily unavailable')
    try:
        raw = await rdb.getdel(_oauth_state_key(state))
    except Exception as exc:
        raise envelope.new_error(503, 'OAUTH_STATE_UNAVAILABLE', 'Authentication service is temporarily unavailable') from exc
    if raw is None:
        raise _google_auth_failed()
    try:
        data = json.loads(raw)
    except (TypeError, ValueError) as exc:
        raise _google_auth_failed() from exc
    nonce, code_verifier = data.get('nonce'), data.get('code_verifier')
    if not isinstance(nonce, str) or not isinstance(code_verifier, str) or not nonce or not code_verifier:
        raise _google_auth_failed()
    return {'nonce': nonce, 'code_verifier': code_verifier}


async def verify_google_id_token(raw_id_token: str, nonce: str | None = None) -> GoogleClaims:
    try:
        info = await asyncio.to_thread(
            google_id_token.verify_oauth2_token, raw_id_token, _google_request
        )
    except ValueError as exc:
        raise _google_auth_failed() from exc
    sub = info.get('sub', '')
    iss = info.get('iss', '')
    if not sub or iss not in ('accounts.google.com', 'https://accounts.google.com'):
        raise _google_auth_failed()
    if not _google_audience_allowed(info.get('aud', '')):
        raise _google_auth_failed()
    if nonce is not None and not hmac.compare_digest(str(info.get('nonce', '')), nonce):
        raise _google_auth_failed()
    return GoogleClaims(
        sub=sub,
        email=info.get('email', ''),
        email_verified=info.get('email_verified') is True or info.get('email_verified') == 'true',
    )


async def _exchange_google_code(
    client_id: str, code: str, code_verifier: str, nonce: str
) -> GoogleClaims:
    form = {
        'code': code,
        'client_id': client_id,
        'client_secret': os.environ.get('GOOGLE_CLIENT_SECRET', ''),
        'redirect_uri': _google_redirect_url(),
        'grant_type': 'authorization_code',
        'code_verifier': code_verifier,
    }
    try:
        token_resp = await _http().post(GOOGLE_TOKEN_URL, data=form)
    except httpx.HTTPError as exc:
        raise _google_auth_failed() from exc
    if token_resp.status_code != 200:
        raise _google_auth_failed()
    try:
        id_token = token_resp.json().get('id_token', '')
    except ValueError as exc:
        raise _google_auth_failed() from exc
    if not id_token:
        raise _google_auth_failed()
    return await verify_google_id_token(id_token, nonce)


def _google_audience_allowed(aud: str) -> bool:
    if not aud:
        return False
    if aud == _google_client_id():
        return True
    for allowed in os.environ.get('GOOGLE_MOBILE_CLIENT_IDS', '').split(','):
        if allowed.strip() and allowed.strip() == aud:
            return True
    return False


def _google_username(email: str) -> str:
    local = email.split('@', 1)[0]
    name = ''.join(c for c in local if c == '_' or c.isascii() and c.isalnum())[:12] or 'user'
    return f"{name}_{runtime.new_uuid().replace('-', '')[:6]}"


async def find_or_create_google_user(pool: AsyncConnectionPool, claims: GoogleClaims) -> runtime.User:
    try:
        return await _find_or_create_google_user_once(pool, claims)
    except pg_errors.UniqueViolation:
        # Retry once: covers a concurrent first login and username-suffix collisions.
        return await _find_or_create_google_user_once(pool, claims)


async def _find_or_create_google_user_once(pool: AsyncConnectionPool, claims: GoogleClaims) -> runtime.User:
    if not claims.sub:
        raise _google_auth_failed()
    async with pool.connection() as conn:
        cursor = await conn.execute(
            """
            SELECT id, username, email, is_active, role, membership, trial_ends_at, revenuecat_app_user_id
            FROM users
            WHERE google_sub = %s
            LIMIT 1
            """,
            (claims.sub,),
        )
        user = runtime._scan_user(await cursor.fetchone())

        if user is None and claims.email and claims.email_verified:
            cursor = await conn.execute(
                """
                SELECT 1 FROM users
                WHERE lower(email) = lower(%s)
                LIMIT 1
                """,
                (claims.email,),
            )
            if await cursor.fetchone() is not None:
                raise envelope.new_error(
                    409, 'ACCOUNT_LINK_REQUIRED',
                    'Sign in with the existing account before linking Google',
                )

        if user is None:
            if not claims.email or not claims.email_verified:
                raise envelope.new_error(
                    403, 'GOOGLE_EMAIL_UNVERIFIED', 'Google account email must be verified'
                )
            cursor = await conn.execute(
                """
                INSERT INTO users (username, email, google_sub, role, membership)
                VALUES (%s, %s, %s, 'user', 'free')
                RETURNING id, username, email, is_active, role, membership, trial_ends_at, revenuecat_app_user_id
                """,
                (_google_username(claims.email), claims.email, claims.sub),
            )
            user = runtime._scan_user(await cursor.fetchone())

    if user is None:
        raise _google_auth_failed()
    if not user.is_active:
        raise envelope.new_error(403, 'USER_INACTIVE', 'User account is disabled')
    await runtime._cache_user(user)
    return user
